### ISSUE TITLE
Unions of typeddicts are resolved too early

### ISSUE THREAD

**Issue by libre-man** [python/mypy#9580]
**Bug Report**

When a variable iis a union of two typeddicts getting an key for that dict is resolved directly as a Union, but this makes it impossible to have dependencies between keys in the dicts.

**To Reproduce**
```
from typing_extensions import TypedDict
import typing as t

class Int(TypedDict):
    a: int
    b: int

class Str(TypedDict):
    a: str
    b: str

opt: t.Union[Int, Str]
opt['a'] + opt['b']
```

**Expected Behavior**

This should type check, as `a` and `b` are always of the same type.

**Actual Behavior**

Mypy gives two errors:
```
test.py:14: error: Unsupported operand types for + ("int" and "str")
test.py:14: error: Unsupported operand types for + ("str" and "int")
test.py:14: note: Both left and right operands are unions
```

**Your Environment**

- Mypy version used: 0.790

**Comment by RomanKornev** [comment]
Just doing
```py
a = {'a': 5}
b = {'b': 10}
a |= b
```
Also gives `Unsupported left operand type for | ("Dict[str, int]")`
PEP 584

**Comment by hauntsaninja** [comment]
@RomanKornev your issue is different from libre-man's and will be fixed by https://github.com/python/typeshed/pull/4671. Note that mypy doesn't yet officially support Python 3.9

**Comment by aalvrz** [comment]
> @RomanKornev Thanks for the report! Your issue is different from libre-man's and will be fixed by [python/typeshed#4671](https://github.com/python/typeshed/pull/4671). Note that mypy doesn't yet officially support Python 3.9

@hauntsaninja Has this been fixed in mypy yet? If so could you please point to the earliest fix version?

**Comment by JelleZijlstra** [comment]
@aalvrz it's been fixed at least since 0.800 but I haven't looked for the exact version. You can try it out in the playground (https://mypy-play.net/?mypy=0.800&python=3.10).

Also, I'm going to close this issue as too advanced for a type checker to understand. It's not specific to TypedDicts; you'll get the same issue if you just use regular classes.

Please open a new issue for any unrelated problems.
